armory/scenarios/poisoning_gtsrb_clbd.py

```diff
@@ -0,0 +1,284 @@
+"""
+Classifier evaluation within ARMORY
+
+Scenario Contributor: MITRE Corporation
+"""
+
+import logging
+from typing import Optional
+import os
+import random
+from copy import deepcopy
+
+import numpy as np
+from tensorflow import set_random_seed, ConfigProto, Session
+from tensorflow.keras.backend import set_session
+from tensorflow.keras.utils import to_categorical
+from tqdm import tqdm
+from PIL import ImageOps, Image
+
+from armory.utils.config_loading import (
+    load_dataset,
+    load_model,
+    load,
+    load_fn,
+)
+from armory.utils import metrics
+from armory.scenarios.base import Scenario
+
+logger = logging.getLogger(__name__)
+
+
+def poison_scenario_preprocessing(batch):
+    img_size = 48
+    img_out = []
+    quantization = 255.0
+    for im in batch:
+        img_eq = ImageOps.equalize(Image.fromarray(im))
+        width, height = img_eq.size
+        min_side = min(img_eq.size)
+        center = width // 2, height // 2
+
+        left = center[0] - min_side // 2
+        top = center[1] - min_side // 2
+        right = center[0] + min_side // 2
+        bottom = center[1] + min_side // 2
+
+        img_eq = img_eq.crop((left, top, right, bottom))
+        img_eq = np.array(img_eq.resize([img_size, img_size])) / quantization
+
+        img_out.append(img_eq)
+
+    return np.array(img_out, dtype=np.float32)
+
+
+def poison_dataset(src_imgs, src_lbls, src, tgt, ds_size, attack, poisoned_indices):
+    # In this example, all images of "src" class have a trigger
+    # added and re-labeled as "tgt" class
+    poison_x = []
+    poison_y = []
+    for idx in range(ds_size):
+        if src_lbls[idx] == src and idx in poisoned_indices:
+            src_img = src_imgs[idx]
+            p_img, p_label = attack.poison(src_img, [tgt])
+            poison_x.append(p_img)
+            poison_y.append(p_label)
+        else:
+            poison_x.append(src_imgs[idx])
+            poison_y.append(src_lbls[idx])
+    poison_x, poison_y = np.array(poison_x), np.array(poison_y)
+
+    return poison_x, poison_y
```

Comment:
Take note that for this attack, we are only poisoning images in the target class. The `attack.poison` method takes the entire dataset and selects the poisoned data within the dataset, and returns the new set containing poisoned data points.

Reply:
Understood. The poison_dataset wrapper is only used with the backdoor object of type PoisoningAttackBackdoor, not the PoisoningAttackCleanLabelBackdoor object.

Reply:
I.e. it is used at eval time, not to poison the training data.